### DISCUSSION TITLE
[React Native] How to pause polling without re-render when a React Native screen gets out of focus

### DISCUSSION THREAD

**Discussion by mk-nickyang** [TanStack/query#6002] [Ideas] [answered]
Hi there, we have a couple screens in our React Native app that poll data regularly, like
```
useQuery({
  queryKey: ['todos'],
  queryFn: async () => {
    const res = await axios.get('/api/data')
    return res.data
  },
  // Refetch the data every 30 seconds
  refetchInterval: 30_000,
})
```
and we would like to temporarily pause it when the screen gets out of focus, to avoid unnecessary network calls and screen updates (_context: In React Native with `react-navigation`,  when user navigates from screen A to screen B, screen A will remain mounted on the stack, it won't unmount, see more [here](https://reactnavigation.org/docs/navigation-lifecycle)_).

To achieve this, we can use `enabled` option + the [`useIsFocused`](https://reactnavigation.org/docs/use-is-focused/) hook , like
```
const isFocused = useIsFocused();

useQuery({
  queryKey: ['todos'],
  queryFn: async () => {
    const res = await axios.get('/api/data')
    return res.data
  },
  enabled: isFocused,
  // Refetch the data every 30 seconds
  refetchInterval: 30_000,
})
```

It works great, but the problem is on the performance side. As @ferrannp well explained in https://github.com/TanStack/query/discussions/5588#discussioncomment-6230198,
> The problem is that `useIsFocused` gets triggered when you blur a screen (and focus), causing a re-render that is very heavy on React Native (nothing should happen on navigation). That is why there is a warning in the docs:
> > Note that using this hook triggers a re-render for the screen when it changes focus. This might cause lags during the animation if your screen is heavy

We thought about using [`useFocusEffect`](https://reactnavigation.org/docs/use-focus-effect/) to tackle this issue, but unfortunately there is no function like `disable` or `enable` returned from `useQuery` where we can potentially do
```
const { enable, disable } = useQuery({
  queryKey: ['todos'],
  queryFn: async () => {
    const res = await axios.get('/api/data')
    return res.data
  },
  // Refetch the data every 30 seconds
  refetchInterval: 30_000,
})

useFocusEffect(
  React.useCallback(() => {
    // Enable query when screen gets in focus
    enable();

    // Disable query when screen goes out of focus
    return () => disable();
  }, [])
);
```

So, I was just wondering if anyone has any ideas on how to pause query polling without causing a re-render when a React Native screen gets out of focus? Thanks in advance 🙏

**Comment by TkDodo** [comment]
isn't that the reason we allowed `notifyOnChangeProps` to take a function?

- https://github.com/TanStack/query/discussions/5588
- https://github.com/TanStack/query/pull/5734

?

**Reply by mk-nickyang** [reply]
Correct me if I'm wrong, but `notifyOnChangeProps` doesn't really stop the polling (network request calls) right? It only stops the re-renders from the query hook

**Reply by TkDodo** [reply]
yes, if you want to stop the polling, you can set `refetchInterval` as a function

**Reply by mk-nickyang** [reply]
Sorry for the late reply.

Yes you are right, we can stop polling if we do something like
```
refetchInterval: () => navigation.isFocused() ? 30_000: false,
```
but once the polling is stopped, doesn't look like it can be resumed when we come back to the screen, because `refetchInterval` can't be re-computed without a re-render or data update i assume?

I was trying to make an expo reproduction, but I keep getting a `Unable to resolve module 'module://react-dom.js'` error for some reason, so I just made an example in web to mimic the React Native issue, please see the link below
https://codesandbox.io/s/rn-polling-demo-3dknf8?file=/src/App.js

Basically the polling can be stopped once we go to the 'Settings' tab, but can't be resumed when we come back to the 'Home' tab

**Reply by TkDodo** [reply]
yeah you need to somehow trigger one re-render wen you come back to the screen. the [docs here](https://tanstack.com/query/v4/docs/react/react-native#disable-re-renders-on-out-of-focus-screens) say:

> This custom hook provides a notifyOnChangeProps option that will return an empty array whenever a screen goes out of focus - effectively stopping any re-renders on that scenario. Whenever the screens gets in focus again, the behavior goes back to normal.

I think you'd want that + the `refetchInterval` function?

**Reply by mk-nickyang** [reply]
Sorry for another late reply, could you please help me understand how `notifyOnChangeProps` can help with this case?

Like if we go with the `refetchInterval` function approach
```
refetchInterval: () => navigation.isFocused() ? 30_000: false
```
as you said, we need to trigger a re-render when we come back to the screen, otherwise `refetchInterval` can't be re-computed. But from my understanding, `notifyOnChangeProps` can only stop the re-render coming from `useQuery`, it doesn't change the fact that we still need to force a re-render right

**Reply by TkDodo** [reply]
sorry, I'm not really an expert on react-native. I don't know when something triggers a re-render and when not. Does `useFocusEvent` not trigger a re-render when the component gets focused ?

**Reply by mk-nickyang** [reply]
I assume you mean `useFocusEffect`? That's correct, [`useFocusEffect`](https://reactnavigation.org/docs/use-focus-effect/) doesn't trigger a re-render when screen focus changes.

The [`useIsFocused`](https://reactnavigation.org/docs/use-is-focused/) hook is the one who does trigger re-render. Here's the React Navigation [documentation page](https://reactnavigation.org/docs/function-after-focusing-screen/#re-rendering-screen-with-the-useisfocused-hook) comparing `focus` event listener, `useFocusEffect` and `useIsFocused` if you are interested

> The `useIsFocused` hook will cause our component to re-render when we focus and unfocus a screen. Using this hook component may introduce unnecessary component re-renders as a screen comes in and out of focus. This could cause issues depending on the type of action we're calling on focusing. Hence we recommend to use this hook only if you need to trigger a re-render. For side-effects such as subscribing to events or fetching data, use the methods described above (`focus` event listener and `useFocusEffect`).

**Reply by mk-nickyang** [reply]
> `useFocusEffect` doesn't trigger a re-render when screen focus changes

This is also where the idea from the original post above comes from
> We thought about using [useFocusEffect](https://reactnavigation.org/docs/use-focus-effect/) to tackle this issue, but unfortunately there is no function like disable or enable returned from useQuery where we can potentially do
```
const { enable, disable } = useQuery({
  queryKey: ['todos'],
  queryFn: async () => {
    const res = await axios.get('/api/data')
    return res.data
  },
  // Refetch the data every 30 seconds
  refetchInterval: 30_000,
})

useFocusEffect(
  React.useCallback(() => {
    // Enable query when screen gets in focus
    enable();

    // Disable query when screen goes out of focus
    return () => disable();
  }, [])
);
```

If it's doable, we can disable/enable the query/polling without any re-render

**Reply by mk-nickyang** [reply]
Hi @TkDodo, just want to circle back to this one to see if you have any new ideas 🙏.

I made an example in web earlier to mimic this React Native issue, please see the link below
https://codesandbox.io/s/rn-polling-demo-cwm82r?file=/src/App.js

Basically we are looking for a way to pause & resume the polling without causing `<HomeTab>` re-render.

We tried to set `refetchInterval` as a function, it does stop the polling, but the polling can't be resumed when we come back to the 'Home' tab, see [here](https://codesandbox.io/s/rn-polling-demo-set-refetchinterval-as-a-function-3dknf8).

So we are running out of ideas 😅. Or do you think re-render is a MUST to achieve this?

**Reply by TkDodo** [reply, marked as answer]
> but the polling can't be resumed when we come back to the 'Home' tab

hmm, have you tried calling `refetchQueries` on the query that should poll manually once when you go back to the `home` tab? I'm thinking that this should re-trigger the interval function, and it works in this forked sandbox:

https://codesandbox.io/s/rn-polling-demo-forked-yzvwzk?file=/src/App.js

Another idea I had: polling stops if the app is seen as not focussed, unless you turn on `refetchIntervalInBackground: true`. This is determined by the global `focusManager`. Now you can set your own focus event handling there and set it to `focused: false` manually if you switch away to a different tab where no polling should occur. Not sure if that's helpful / doable, but it's a possibility. See the forked sandbox here: 

https://codesandbox.io/s/rn-polling-demo-forked-hd5pfm?file=/src/App.js

**Reply by mk-nickyang** [reply]
Appreciate for your reply 🙏 Just tried both ideas

> calling `refetchQueries` when go back to the home tab

Think this does work, for some reason I never thought using refetch to re-trigger the interval function 😅 Brilliant idea!

The only downside I can think of is just when user constantly switches between tabs, query will be refetched a lot. But I guess we can check the diff between `Date.now() - dataUpdatedAt` and `refetchInterval`, to force this manual refetch to also respect `refetchInterval`? What do you think 🤔

> use `focusManager`

This one also works for the example above. But i guess since `focusManager` is a global thing instead of on the observer level, when we set `focused: false`, it will also stop other queries polling from other tabs right? (assuming there are other polling queries)

**Reply by TkDodo** [reply]
> to force this manual refetch to also respect refetchInterval

I think what would happen is that since the page doesn't re-render, the interval doesn't re-start if you don't refetch. You need some "signal" to resume the interval: either a re-render of the component, or a refetch.

>  it will also stop other queries polling from other tabs right?

yes, it would.

**Reply by mk-nickyang** [reply]
Yeah right, make sense. Thanks again for all the help so far, will close this discussion now 🙏
